ipn/ipnauth/accesschk.go

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package ipnauth
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"tailscale.com/ipn"
+	"tailscale.com/tailcfg"
+)
+
+// errNotAllowed is an error returned when access is neither explicitly allowed,
+// nor denied with a more specific error.
+var errNotAllowed = ipn.NewAccessDeniedError("the requested operation is not allowed")
+
+// AccessCheckResult represents the result of an access check.
+// Its zero value is valid and indicates that the access request
+// has neither been explicitly allowed nor denied for a specific reason.
+// In this case, access will be denied, returning errNotAllowed.
+type AccessCheckResult struct {
+	err       error
+	hasResult bool
+}
+
+// AllowAccess returns a new AccessCheckResult indicating that
+// the requested access has been allowed.
+func AllowAccess() AccessCheckResult {
+	return AccessCheckResult{hasResult: true}
+}
+
+// DenyAccess returns a new AccessCheckResult indicating that
+// the requested access has been denied with the specified non-nil err.
+// It panics if err is nil.
```

Comment:
It's not necessary in the sense that returning an error instead of panicking won't lead to data corruption or privilege escalation. OTOH, a programming error in the access checking code, most likely within `ipnauth` itself, seems bad enough that we'd want to know about it sooner rather than later; panicking also provides additional information, such as the call stack.

Happy to change if we prefer returning errors in cases like this, but would it change your mind if we made `DenyAccess` and friends non-exported?